src/conversions/num_rational.rs

```diff
@@ -0,0 +1,228 @@
+//! num-rational = "0.4.1"
+//! ```
+//!
+//! # Example
+//!
+//! Rust code to create a function that adds five to a fraction:
+//!
+//! ```rust
+//! use num_rational::Ratio;
+//! use pyo3::prelude::*;
+//!
+//! #[pyfunction]
+//! fn add_five_to_fraction(fraction: Ratio<i32>) -> Ratio<i32> {
+//!     fraction + Ratio::new(5, 1)
+//! }
+//!
+//! #[pymodule]
+//! fn my_module(m: &Bound<'_, PyModule>) -> PyResult<()> {
+//!     m.add_function(wrap_pyfunction!(add_five_to_fraction, m)?)?;
+//!     Ok(())
+//! }
+//! ```
+//!
+//! Python code that validates the functionality:
+//! ```python
+//! from my_module import add_five_to_fraction
+//! from fractions import Fraction
+//!
+//! fraction = Fraction(2,1)
+//! fraction_plus_five = add_five_to_fraction(f)
+//! assert fraction + 5 == fraction_plus_five
+//! ```
+
+use crate::ffi;
+use crate::sync::GILOnceCell;
+use crate::types::any::PyAnyMethods;
+use crate::types::PyLong;
+use crate::types::PyType;
+use crate::{Bound, FromPyObject, IntoPy, Py, PyAny, PyObject, PyResult, Python, ToPyObject};
+use std::os::raw::c_char;
+
+#[cfg(feature = "num-bigint")]
+use num_bigint::BigInt;
+use num_rational::Ratio;
+
+static FRACTION_CLS: GILOnceCell<Py<PyType>> = GILOnceCell::new();
+
+fn get_fraction_cls(py: Python<'_>) -> PyResult<&Bound<'_, PyType>> {
+    FRACTION_CLS.get_or_try_init_type_ref(py, "fractions", "Fraction")
+}
+
+macro_rules! rational_conversion {
+    ($int: ty) => {
+        impl<'py> FromPyObject<'py> for Ratio<$int> {
+            fn extract_bound(obj: &Bound<'py, PyAny>) -> PyResult<Self> {
+                let py = obj.py();
+                let py_numerator_obj = unsafe {
+                    ffi::PyObject_GetAttrString(
+                        obj.as_ptr(),
+                        "numerator\0".as_ptr() as *const c_char,
+                    )
+                };
+                let py_denominator_obj = unsafe {
+                    ffi::PyObject_GetAttrString(
+                        obj.as_ptr(),
+                        "denominator\0".as_ptr() as *const c_char,
+                    )
+                };
+                let numerator_owned: Py<PyLong> =
+                    unsafe { Py::from_owned_ptr_or_err(py, ffi::PyNumber_Long(py_numerator_obj))? };
+                let denominator_owned: Py<PyLong> = unsafe {
+                    Py::from_owned_ptr_or_err(py, ffi::PyNumber_Long(py_denominator_obj))?
+                };
+                let rs_numerator: $int = numerator_owned.bind(py).extract()?;
+                let rs_denominator: $int = denominator_owned.bind(py).extract()?;
+                Ok(Ratio::new(rs_numerator, rs_denominator))
+            }
+        }
+
+        impl ToPyObject for Ratio<$int> {
+            fn to_object(&self, py: Python<'_>) -> PyObject {
+                let fraction_cls = get_fraction_cls(py).expect("failed to load fractions.Fraction");
+                let ret = fraction_cls
+                    .call1((self.to_string(),))
```

Comment:
Is there a reason not to use `numer()` and `denom()` here? That might be a bit more efficient than allocating a `String`.